File: sdk/core/unity/Runtime/Core/FallbackWaterfall.cs
using System;
using System.Collections.Generic;
using System.Threading;
using System.Threading.Tasks;

namespace RivalApex.SDK.Core
{
    /// <summary>
    /// FallbackWaterfall - Implements a waterfall mediation strategy for ad loading.
    ///
    /// When one ad source fails, it automatically falls back to the next source
    /// in the waterfall. This ensures maximum fill rate by trying multiple
    /// demand partners in priority order.
    /// </summary>
    public class FallbackWaterfall<T>
    {
        /// <summary>
        /// Default timeout for source loading (ms)
        /// </summary>
        public int DefaultTimeoutMs { get; }
        
        public FallbackWaterfall(int defaultTimeoutMs = 5000)
        {
            DefaultTimeoutMs = defaultTimeoutMs;
        }
        
        /// <summary>
        /// Represents an ad source in the waterfall
        /// </summary>
        public class WaterfallSource
        {
            public string Id { get; }
            public int Priority { get; }
            public int TimeoutMs { get; }
            public Func<CancellationToken, Task<WaterfallResult<T>>> Loader { get; }
            
            public WaterfallSource(
                string id,
                int priority,
                int timeoutMs,
                Func<CancellationToken, Task<WaterfallResult<T>>> loader)
            {
                Id = id;
                Priority = priority;
                TimeoutMs = timeoutMs;
                Loader = loader;
            }
        }
        
        /// <summary>
        /// Result of attempting to load from a source
        /// </summary>
        public abstract class WaterfallResult<TResult>
        {
            public class Success : WaterfallResult<TResult>
            {
                public TResult Data { get; }
                public Success(TResult data) => Data = data;
            }
            
            public class NoFill : WaterfallResult<TResult>
            {
                public string Reason { get; }
                public NoFill(string reason = "No ad available") => Reason = reason;
            }
            
            public class Error : WaterfallResult<TResult>
            {
                public Exception Exception { get; }
                public Error(Exception exception) => Exception = exception;
            }
            
            public class Timeout : WaterfallResult<TResult> { }
        }
        
        public enum AttemptResultType
        {
            Success, NoFill, Error, Timeout, Skipped
        }
        
        /// <summary>
        /// Details about each attempt in the waterfall
        /// </summary>
        public readonly struct AttemptDetail
        {
            public string SourceId { get; }
            public int Priority { get; }
            public long DurationMs { get; }
            public AttemptResultType Result { get; }
            
            public AttemptDetail(string sourceId, int priority, long durationMs, AttemptResultType result)
            {
                SourceId = sourceId;
                Priority = priority;
                DurationMs = durationMs;
                Result = result;
            }
        }
        
        /// <summary>
        /// Waterfall execution result with metadata
        /// </summary>
        public class ExecutionResult
        {
            public WaterfallResult<T> Result { get; }
            public string SourceId { get; }
            public int AttemptsCount { get; }
            public long TotalDurationMs { get; }
            public IReadOnlyList<AttemptDetail> AttemptDetails { get; }
            
            public ExecutionResult(
                WaterfallResult<T> result,
                string sourceId,
                int attemptsCount,
                long totalDurationMs,
                IReadOnlyList<AttemptDetail> attemptDetails)
            {
                Result = result;
                SourceId = sourceId;
                AttemptsCount = attemptsCount;
                TotalDurationMs = totalDurationMs;
                AttemptDetails = attemptDetails;
            }
        }
        
        /// <summary>
        /// Performance statistics for a source
        /// </summary>
        public class SourceStats
        {
            private int _successCount;
            private int _failureCount;
            private int _timeoutCount;
            private int _noFillCount;
            private long _totalLatencyMs;
            
            public int SuccessCount => _successCount;
            public int FailureCount => _failureCount;
            public int TimeoutCount => _timeoutCount;
            public int NoFillCount => _noFillCount;
            
            public int TotalAttempts => _successCount + _failureCount + _timeoutCount + _noFillCount;
            
            public float SuccessRate => TotalAttempts > 0 ? (float)_successCount / TotalAttempts : 0f;
            
            public long AverageLatencyMs => TotalAttempts > 0 ? _totalLatencyMs / TotalAttempts : 0;
            
            public void RecordSuccess(long latencyMs)
            {
                Interlocked.Increment(ref _successCount);
                Interlocked.Add(ref _totalLatencyMs, latencyMs);
            }
            
            public void RecordFailure(long latencyMs)
            {
                Interlocked.Increment(ref _failureCount);
                Interlocked.Add(ref _totalLatencyMs, latencyMs);
            }
            
            public void RecordTimeout(long latencyMs)
            {
                Interlocked.Increment(ref _timeoutCount);
                Interlocked.Add(ref _totalLatencyMs, latencyMs);
            }
            
            public void RecordNoFill(long latencyMs)
            {
                Interlocked.Increment(ref _noFillCount);
                Interlocked.Add(ref _totalLatencyMs, latencyMs);
            }
        }
        
        private readonly Dictionary<string, SourceStats> _sourceStats = new();
        private readonly object _statsLock = new();
        
        /// <summary>
        /// Executes the waterfall, trying each source in priority order
        /// </summary>
        public async Task<ExecutionResult> ExecuteAsync(
            IEnumerable<WaterfallSource> sources,
            CancellationToken cancellationToken = default)
        {
            var startTime = DateTime.UtcNow;
            var attemptDetails = new List<AttemptDetail>();
            var sortedSources = new List<WaterfallSource>(sources);
            sortedSources.Sort((a, b) => a.Priority.CompareTo(b.Priority));
            
            foreach (var source in sortedSources)
            {
                cancellationToken.ThrowIfCancellationRequested();
                
                var attemptStart = DateTime.UtcNow;
                WaterfallResult<T> result;
                
                try
                {
                    using var timeoutCts = new CancellationTokenSource(source.TimeoutMs);
                    using var linkedCts = CancellationTokenSource.CreateLinkedTokenSource(
                        cancellationToken, timeoutCts.Token);
                    
                    result = await source.Loader(linkedCts.Token);
                }
                catch (OperationCanceledException) when (!cancellationToken.IsCancellationRequested)
                {
                    result = new WaterfallResult<T>.Timeout();
                }
                catch (Exception ex)
                {
                    result = new WaterfallResult<T>.Error(ex);
                }
                
                var durationMs = (long)(DateTime.UtcNow - attemptStart).TotalMilliseconds;
                RecordStats(source.Id, result, durationMs);
                
                var attemptResultType = result switch
                {
                    WaterfallResult<T>.Success => AttemptResultType.Success,
                    WaterfallResult<T>.NoFill => AttemptResultType.NoFill,
                    WaterfallResult<T>.Error => AttemptResultType.Error,
                    WaterfallResult<T>.Timeout => AttemptResultType.Timeout,
                    _ => AttemptResultType.Error
                };
                
                attemptDetails.Add(new AttemptDetail(
                    source.Id, source.Priority, durationMs, attemptResultType));
                
                if (result is WaterfallResult<T>.Success)
                {
                    return new ExecutionResult(
                        result,
                        source.Id,
                        attemptDetails.Count,
                        (long)(DateTime.UtcNow - startTime).TotalMilliseconds,
                        attemptDetails);
                }
            }
            
            // All sources exhausted
            return new ExecutionResult(
                new WaterfallResult<T>.NoFill($"All {sortedSources.Count} sources exhausted"),
                sortedSources.Count > 0 ? sortedSources[^1].Id : "none",
                attemptDetails.Count,
                (long)(DateTime.UtcNow - startTime).TotalMilliseconds,
                attemptDetails);
        }
        
        public SourceStats GetStats(string sourceId)
        {
            lock (_statsLock)
            {
                return _sourceStats.TryGetValue(sourceId, out var stats) ? stats : null;
            }
        }
        
        public Dictionary<string, SourceStats> GetAllStats()
        {
            lock (_statsLock)
            {
                return new Dictionary<string, SourceStats>(_sourceStats);
            }
        }
        
        public void ClearStats()
        {
            lock (_statsLock)
            {
                _sourceStats.Clear();
            }
        }
        
        public WaterfallSource CreateSource(
            string id,
            int priority,
            int? timeoutMs = null,
            Func<CancellationToken, Task<WaterfallResult<T>>> loader)
        {
            return new WaterfallSource(id, priority, timeoutMs ?? DefaultTimeoutMs, loader);
        }
        
        private void RecordStats(string sourceId, WaterfallResult<T> result, long durationMs)
        {
            SourceStats stats;
            lock (_statsLock)
            {
                if (!_sourceStats.TryGetValue(sourceId, out stats))
                {
                    stats = new SourceStats();
                    _sourceStats[sourceId] = stats;
                }
            }
            
            switch (result)
            {
                case WaterfallResult<T>.Success:
                    stats.RecordSuccess(durationMs);
                    break;
                case WaterfallResult<T>.NoFill:
                    stats.RecordNoFill(durationMs);
                    break;
                case WaterfallResult<T>.Error:
                    stats.RecordFailure(durationMs);
                    break;
                case WaterfallResult<T>.Timeout:
                    stats.RecordTimeout(durationMs);
                    break;
            }
        }
    }
}

File: sdk/web/src/fallbackWaterfall.ts
/**
 * FallbackWaterfall - Implements a waterfall mediation strategy for ad loading.
 *
 * When one ad source fails, it automatically falls back to the next source
 * in the waterfall. This ensures maximum fill rate by trying multiple
 * demand partners in priority order.
 */

/** Default timeout for source loading (ms) */
export const DEFAULT_TIMEOUT_MS = 5000;

/** Result type for attempts */
export type AttemptResultType = 'success' | 'noFill' | 'error' | 'timeout' | 'skipped';

/** Result of attempting to load from a source */
export type WaterfallResult<T> =
  | { type: 'success'; data: T }
  | { type: 'noFill'; reason: string }
  | { type: 'error'; error: Error }
  | { type: 'timeout' };

/** Represents an ad source in the waterfall */
export interface WaterfallSource<T> {
  readonly id: string;
  readonly priority: number;
  readonly timeoutMs: number;
  readonly loader: (signal: AbortSignal) => Promise<WaterfallResult<T>>;
}

/** Details about each attempt in the waterfall */
export interface AttemptDetail {
  readonly sourceId: string;
  readonly priority: number;
  readonly durationMs: number;
  readonly result: AttemptResultType;
}

/** Waterfall execution result with metadata */
export interface ExecutionResult<T> {
  readonly result: WaterfallResult<T>;
  readonly sourceId: string;
  readonly attemptsCount: number;
  readonly totalDurationMs: number;
  readonly attemptDetails: readonly AttemptDetail[];
}

/** Performance statistics for a source */
export interface SourceStats {
  successCount: number;
  failureCount: number;
  timeoutCount: number;
  noFillCount: number;
  totalLatencyMs: number;
  readonly totalAttempts: number;
  readonly successRate: number;
  readonly averageLatencyMs: number;
}

function createSourceStats(): SourceStats {
  return {
    successCount: 0,
    failureCount: 0,
    timeoutCount: 0,
    noFillCount: 0,
    totalLatencyMs: 0,
    get totalAttempts() {
      return this.successCount + this.failureCount + this.timeoutCount + this.noFillCount;
    },
    get successRate() {
      return this.totalAttempts > 0 ? this.successCount / this.totalAttempts : 0;
    },
    get averageLatencyMs() {
      return this.totalAttempts > 0 ? this.totalLatencyMs / this.totalAttempts : 0;
    }
  };
}

/** FallbackWaterfall class for managing waterfall execution */
export class FallbackWaterfall<T> {
  private readonly defaultTimeoutMs: number;
  private readonly sourceStats = new Map<string, SourceStats>();
  
  constructor(defaultTimeoutMs: number = DEFAULT_TIMEOUT_MS) {
    this.defaultTimeoutMs = defaultTimeoutMs;
  }
  
  /**
   * Executes the waterfall, trying each source in priority order until
   * one succeeds or all sources are exhausted.
   */
  async execute(
    sources: readonly WaterfallSource<T>[],
    signal?: AbortSignal
  ): Promise<ExecutionResult<T>> {
    const startTime = Date.now();
    const attemptDetails: AttemptDetail[] = [];
    const sortedSources = [...sources].sort((a, b) => a.priority - b.priority);
    
    for (const source of sortedSources) {
      // Check for cancellation
      if (signal?.aborted) {
        throw new DOMException('Waterfall execution aborted', 'AbortError');
      }
      
      const attemptStart = Date.now();
      let result: WaterfallResult<T>;
      
      try {
        result = await this.executeWithTimeout(source, signal);
      } catch (error) {
        if (error instanceof DOMException && error.name === 'AbortError') {
          result = { type: 'timeout' };
        } else {
          result = { type: 'error', error: error instanceof Error ? error : new Error(String(error)) };
        }
      }
      
      const durationMs = Date.now() - attemptStart;
      this.recordStats(source.id, result, durationMs);
      
      const attemptResultType = this.resultToType(result);
      attemptDetails.push({
        sourceId: source.id,
        priority: source.priority,
        durationMs,
        result: attemptResultType
      });
      
      if (result.type === 'success') {
        return {
          result,
          sourceId: source.id,
          attemptsCount: attemptDetails.length,
          totalDurationMs: Date.now() - startTime,
          attemptDetails
        };
      }
    }
    
    // All sources exhausted
    return {
      result: { type: 'noFill', reason: `All ${sortedSources.length} sources exhausted` },
      sourceId: sortedSources[sortedSources.length - 1]?.id ?? 'none',
      attemptsCount: attemptDetails.length,
      totalDurationMs: Date.now() - startTime,
      attemptDetails
    };
  }
  
  /**
   * Executes waterfall with parallel preloading of lower-priority sources.
   */
  async executeWithParallelPreload(
    sources: readonly WaterfallSource<T>[],
    preloadCount: number = 1,
    signal?: AbortSignal
  ): Promise<ExecutionResult<T>> {
    const startTime = Date.now();
    const attemptDetails: AttemptDetail[] = [];
    const sortedSources = [...sources].sort((a, b) => a.priority - b.priority);
    
    const preloadResults = new Map<string, Promise<WaterfallResult<T>>>();
    
    for (let i = 0; i < sortedSources.length; i++) {
      const source = sortedSources[i];
      
      if (signal?.aborted) {
        throw new DOMException('Waterfall execution aborted', 'AbortError');
      }
      
      // Start preloading next sources
      for (let j = i + 1; j <= i + preloadCount && j < sortedSources.length; j++) {
        const preloadSource = sortedSources[j];
        if (!preloadResults.has(preloadSource.id)) {
          preloadResults.set(
            preloadSource.id,
            this.executeWithTimeout(preloadSource, signal).catch(err => ({
              type: 'error' as const,
              error: err instanceof Error ? err : new Error(String(err))
            }))
          );
        }
      }
      
      const attemptStart = Date.now();
      let result: WaterfallResult<T>;
      
      // Use preloaded result if available, otherwise execute now
      const preloadedPromise = preloadResults.get(source.id);
      if (preloadedPromise) {
        try {
          result = await preloadedPromise;
        } catch (error) {
          result = { type: 'error', error: error instanceof Error ? error : new Error(String(error)) };
        }
      } else {
        try {
          result = await this.executeWithTimeout(source, signal);
        } catch (error) {
          if (error instanceof DOMException && error.name === 'AbortError') {
            result = { type: 'timeout' };
          } else {
            result = { type: 'error', error: error instanceof Error ? error : new Error(String(error)) };
          }
        }
      }
      
      const durationMs = Date.now() - attemptStart;
      this.recordStats(source.id, result, durationMs);
      
      attemptDetails.push({
        sourceId: source.id,
        priority: source.priority,
        durationMs,
        result: this.resultToType(result)
      });
      
      if (result.type === 'success') {
        return {
          result,
          sourceId: source.id,
          attemptsCount: attemptDetails.length,
          totalDurationMs: Date.now() - startTime,
          attemptDetails
        };
      }
    }
    
    return {
      result: { type: 'noFill', reason: `All ${sortedSources.length} sources exhausted` },
      sourceId: sortedSources[sortedSources.length - 1]?.id ?? 'none',
      attemptsCount: attemptDetails.length,
      totalDurationMs: Date.now() - startTime,
      attemptDetails
    };
  }
  
  /**
   * Gets statistics for a specific source.
   */
  getStats(sourceId: string): SourceStats | undefined {
    return this.sourceStats.get(sourceId);
  }
  
  /**
   * Gets statistics for all sources.
   */
  getAllStats(): Map<string, SourceStats> {
    return new Map(this.sourceStats);
  }
  
  /**
   * Clears all recorded statistics.
   */
  clearStats(): void {
    this.sourceStats.clear();
  }
  
  /**
   * Creates a source with the given parameters.
   */
  createSource(
    id: string,
    priority: number,
    loader: (signal: AbortSignal) => Promise<WaterfallResult<T>>,
    timeoutMs: number = this.defaultTimeoutMs
  ): WaterfallSource<T> {
    return { id, priority, timeoutMs, loader };
  }
  
  private async executeWithTimeout(
    source: WaterfallSource<T>,
    parentSignal?: AbortSignal
  ): Promise<WaterfallResult<T>> {
    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), source.timeoutMs);
    
    // Link to parent signal
    const abortHandler = () => controller.abort();
    parentSignal?.addEventListener('abort', abortHandler);
    
    try {
      return await source.loader(controller.signal);
    } finally {
      clearTimeout(timeoutId);
      parentSignal?.removeEventListener('abort', abortHandler);
    }
  }
  
  private resultToType(result: WaterfallResult<T>): AttemptResultType {
    switch (result.type) {
      case 'success': return 'success';
      case 'noFill': return 'noFill';
      case 'error': return 'error';
      case 'timeout': return 'timeout';
    }
  }
  
  private recordStats(sourceId: string, result: WaterfallResult<T>, durationMs: number): void {
    let stats = this.sourceStats.get(sourceId);
    if (!stats) {
      stats = createSourceStats();
      this.sourceStats.set(sourceId, stats);
    }
    
    stats.totalLatencyMs += durationMs;
    
    switch (result.type) {
      case 'success':
        stats.successCount++;
        break;
      case 'noFill':
        stats.noFillCount++;
        break;
      case 'error':
        stats.failureCount++;
        break;
      case 'timeout':
        stats.timeoutCount++;
        break;
    }
  }
}

/** Creates a new FallbackWaterfall instance */
export function createWaterfall<T>(defaultTimeoutMs?: number): FallbackWaterfall<T> {
  return new FallbackWaterfall<T>(defaultTimeoutMs);
}

/** Helper to create a success result */
export function success<T>(data: T): WaterfallResult<T> {
  return { type: 'success', data };
}

/** Helper to create a no-fill result */
export function noFill<T>(reason: string = 'No ad available'): WaterfallResult<T> {
  return { type: 'noFill', reason };
}

/** Helper to create an error result */
export function error<T>(err: Error): WaterfallResult<T> {
  return { type: 'error', error: err };
}

/** Helper to create a timeout result */
export function timeout<T>(): WaterfallResult<T> {
  return { type: 'timeout' };
}
